File: iOS/Account/Sources/Account/ByEmailAndPassword/EmailAndPasswordSignInViewModel.swift
import Foundation
import Core
import UI
import Session
import ViewingState

public class EmailAndPasswordSignInViewModel: ObservableObject {
    @Published @MainActor private(set) var state: LoadState<AccountBundle> = .idle

    private let signInWithEmailAndPasswordUseCase: SignInWithEmailAndPasswordUseCase

    public init(signInWithEmailAndPasswordUseCase: SignInWithEmailAndPasswordUseCase) {
        self.signInWithEmailAndPasswordUseCase = signInWithEmailAndPasswordUseCase
    }
    
    @MainActor
    func reset() {
        state = .idle
    }
    
    @MainActor
    func signInWithEmailAndPassword(email: String, password: String) {
        Task {
            state = .loading
            
            let result = await signInWithEmailAndPasswordUseCase.execute(email: Email(email)!, password: password)
            switch result {
            case .success(let accountBundle):
                state = .loaded(accountBundle)
            case .failure(let error):
                state = .failure(error.mapOrDefaultLocalized { $0.asEmailAndPasswordSignInLocalizedError() })
            }
        }
    }
}

File: Android/account/src/main/java/com/felipearpa/tyche/account/byemailandpassword/EmailAndPasswordSignInViewModel.kt
package com.felipearpa.tyche.account.byemailandpassword

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.felipearpa.tyche.core.type.Email
import com.felipearpa.tyche.session.AccountBundle
import com.felipearpa.tyche.session.authentication.application.SignInWithEmailAndPassword
import com.felipearpa.tyche.ui.exception.mapOrDefaultLocalized
import com.felipearpa.ui.state.LoadState
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class EmailAndPasswordSignInViewModel(
    private val signInWithEmailAndPassword: SignInWithEmailAndPassword,
) : ViewModel() {
    private val _state =
        MutableStateFlow<LoadState<AccountBundle>>(LoadState.Idle)
    val state = _state.asStateFlow()

    fun reset() {
        _state.value = LoadState.Idle
    }

    fun signInWithEmailAndPassword(email: String, password: String) {
        viewModelScope.launch {
            _state.emit(LoadState.Loading)
            signInWithEmailAndPassword.execute(email = Email(email), password = password)
                .onSuccess { accountBundle -> _state.emit(LoadState.Loaded(accountBundle)) }
                .onFailure { exception ->
                    _state.emit(
                        LoadState.Failure(exception.mapOrDefaultLocalized { it.asEmailAndPasswordSignInLocalized() }),
                    )
                }
        }
    }
}
